### PR TITLE
LightningEstimator: add support for in memory dateloader

### PR DESCRIPTION
**Pull request by chongxiaoc**
## Checklist before submitting

- [ ] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [ ] Did you write any tests to validate this change?  
- [ ] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description

Fixes # (issue).

## Review process to land 

1. All tests and other checks must succeed.
2. At least one member of the [technical steering committee](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md) must review and approve.
3. If any member of the technical steering committee requests changes, they must be addressed.

### CODE REVIEW

File: horovod/spark/data_loaders/pytorch_data_loaders.py [hunk 1]
```diff
@@ -95,14 +96,60 @@ def _iterate(self):
...
+              f"shuffle={shuffle}"
+              f"limit_step_per_epoch={limit_step_per_epoch}")
+
+        self.dataloader = InMemBatchedDataLoader(reader, batch_size=batch_size, num_epochs=num_epochs,
+                                                 rows_capacity=batch_size*limit_step_per_epoch, shuffle=shuffle)
+        self.iterator = iter(self.dataloader)
+
+    def __len__(self):
+        # We cannot infer length from reader.
+        return self.limit_step_per_epoch if self.limit_step_per_epoch != -1 else 0
```
Comment (tgaddair) [inline horovod/spark/data_loaders/pytorch_data_loaders.py:129]: Don't we raise an exception if `self.limit_step_per_epoch == -1`? Can we then just have this always return `self.limit_step_per_epoch`?

File: horovod/spark/lightning/remote.py [hunk 2]
```diff
@@ -267,11 +271,19 @@ def set_data_loader(model, data_path, dataloader_attr, reader_worker_count, read
                             storage_options=store.storage_options,
                             **reader_factory_kwargs) as reader:
             def dataloader_fn():
-                return data_loader_cls(reader=reader, batch_size=batch_size,
-                                       shuffling_queue_capacity=shuffling_queue_capacity,
-                                       name=name,
-                                       limit_step_per_epoch=limit_step_per_epoch,
-                                       verbose=verbose)
+                if inmemory_cache_all:
+                    shuffle = True if shuffling_queue_capacity > 0 else False
```
Comment (tgaddair) [inline horovod/spark/lightning/remote.py:275]: Can be simplified:

```
shuffle = shuffling_queue_capacity > 0
```

File: horovod/spark/lightning/remote.py [hunk 3]
```diff
@@ -267,11 +271,19 @@ def set_data_loader(model, data_path, dataloader_attr, reader_worker_count, read
...
+                    # Use inmem dataloader
+                    return data_loader_cls(reader, batch_size, epochs,
```
Comment (tgaddair) [inline horovod/spark/lightning/remote.py:277]: Looks like these are almost the same except one param. So maybe do:

```
kwargs = dict(
reader=reader, batch_size=batch_size,
                                           shuffling_queue_capacity=shuffling_queue_capacity,
                                           name=name,
                                           limit_step_per_epoch=limit_step_per_epoch,
                                           verbose=verbose
)
if inmemory_cache_all:
    kwargs['shuffle'] = shuffling_queue_capacity > 0
else:
    kwargs['shuffling_queue_capacity'] = shuffling_queue_capacity
```

**Review by tgaddair — COMMENTED** [review]
Looks good! Just couple of small things.

**Comment by chongxiaoc** [comment]
comments addressed.

**Comment by chongxiaoc** [comment]
@tgaddair Can you take one more look?

**Review by tgaddair — APPROVED** [review]
LGTM!

**Pull request merged**